pkg/frontend/querymiddleware/codec_test.go

```diff
@@ -312,16 +321,26 @@ func TestResponseRoundtrip(t *testing.T) {
 			require.NoError(t, err)
 			assert.Equal(t, tc.expected, decoded)
 
+			metrics := gatherMetrics(t, reg)
```

Comment:
I couldn't find a better way to test this, open to feedback - I don't want to use `testutil.GatherAndCompare()` because it only supports comparing all metrics in a histogram, rather than individual metrics like I'm doing here.

Reply:
I don't have a better suggestion, I've also been frustrated by this limitation of `GatherAndCompare`.

Reply:
I confirm you can't use `GatherAndCompare` for unpredictable metric values (like in this case). However, instead of writing `gatherMetrics()` from scratch, I recommend to use our utilities in `pkg/util/metrics_helper.go`.

You can call `NewMetricFamilyMapFromGatherer()`. Then you can insect the metric data inside each `MetricFamily`.

Reply:
Addressed in 21fa311.